token/program-2022/src/processor.rs

```diff
@@ -352,47 +353,71 @@ impl Processor {
         // compute costs, ie:
         // if self_transfer || amount == 0
         check_program_account(source_account_info.owner)?;
-        check_program_account(dest_account_info.owner)?;
 
         // This check MUST occur just before the amounts are manipulated
         // to ensure self-transfers are fully validated
         if self_transfer {
             return Ok(());
         }
 
-        // self-transfer was dealt with earlier, so this *should* be safe
-        let mut dest_account_data = dest_account_info.data.borrow_mut();
-        let mut dest_account = StateWithExtensionsMut::<Account>::unpack(&mut dest_account_data)?;
-
-        if dest_account.base.is_frozen() {
-            return Err(TokenError::AccountFrozen.into());
-        }
-        if !cmp_pubkeys(&source_account.base.mint, &dest_account.base.mint) {
-            return Err(TokenError::MintMismatch.into());
-        }
+        let wsol_source = source_account.base.is_native();
+        let sol_destination = *dest_account_info.owner == system_program::id();
```

Comment:
We should assert that `data.len() = 0` here as well.  `system::transfer` disallows debiting data carrying system accounts (durable nonce mis-feature).

https://github.com/solana-labs/solana/blob/27aaf9df85633b13c32d44d139528d8948141a53/runtime/src/system_instruction_processor.rs#L177-L180